tests/matrix_free/fe_evaluation_shift.cc

```diff
@@ -0,0 +1,433 @@
+      const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+      const unsigned int                                  dof_no  = 0,
+      const unsigned int                                  quad_no = 0,
+      const unsigned int first_selected_component                 = 0,
+      const unsigned int active_fe_index   = numbers::invalid_unsigned_int,
+      const unsigned int active_quad_index = numbers::invalid_unsigned_int)
+      : BaseClass(matrix_free,
+                  dof_no,
+                  quad_no,
+                  first_selected_component,
+                  active_fe_index,
+                  active_quad_index)
+      , shape_info_base(this->data){
+
+        };
+
+    const internal::MatrixFreeFunctions::ShapeInfo<VectorizedArrayType>
+      *shape_info_base;
+
+    void
+    reinit(unsigned int cell_batch_index,
+           const internal::MatrixFreeFunctions::ShapeInfo<VectorizedArrayType>
+             *shape_info = nullptr);
+
+    void
+    reinit(unsigned int cell_batch_index,
+           unsigned int face_number,
+           const internal::MatrixFreeFunctions::ShapeInfo<VectorizedArrayType>
+             *shape_info = nullptr);
+  };
+
+  template <int dim,
+            int fe_degree,
+            int n_q_points_1d,
+            int n_components_,
+            typename Number,
+            typename VectorizedArrayType>
+  inline void
+  FEEvaluationShift<dim,
+                    fe_degree,
+                    n_q_points_1d,
+                    n_components_,
+                    Number,
+                    VectorizedArrayType>::
+    reinit(const unsigned int cell_batch_index,
+           const internal::MatrixFreeFunctions::ShapeInfo<VectorizedArrayType>
+             *shape_info)
+  {
+    Assert(this->mapped_geometry == nullptr,
+           ExcMessage(
+             "FEEvaluation was initialized without a matrix-free object."
+             " Integer indexing is not possible"));
+    if (this->mapped_geometry != nullptr)
+      return;
+
+    Assert(this->dof_info != nullptr, ExcNotInitialized());
+    Assert(this->mapping_data != nullptr, ExcNotInitialized());
+    this->cell = cell_batch_index;
+    this->cell_type =
+      this->matrix_free->get_mapping_info().get_cell_type(cell_batch_index);
+
+    const unsigned int offsets =
+      this->mapping_data->data_index_offsets[cell_batch_index];
+    this->jacobian = &this->mapping_data->jacobians[0][offsets];
+    this->J_value  = &this->mapping_data->JxW_values[offsets];
+
+    for (unsigned int i = 0; i < VectorizedArrayType::size(); ++i)
+      this->cell_ids[i] = cell_batch_index * VectorizedArrayType::size() + i;
+
+    this->interior_face = true;
+
+    if (shape_info != nullptr)
+      {
+        AssertDimension(shape_info->n_q_points, this->data->n_q_points);
+        AssertDimension(shape_info->dofs_per_component_on_cell,
+                        this->data->dofs_per_component_on_cell);
+        this->data = shape_info;
+      }
```

Comment:
You're not using this path of the code, do you? Do you want to extend the test to also make use of a `ShapeInfo` evaluated outside of the cell?

Reply:
Yes this is a good idea!

Reply:
Updated!